kore/src/Kore/Step/Step.hs

```diff
@@ -224,8 +229,26 @@ unifyRule
     let
         RulePattern { requires = ruleRequires } = rule'
         requires' = Predicate.fromPredicate ruleRequires
-    unification' <- Substitution.normalizeExcept (unification <> requires')
+    unification' <- removeRedundancy unification requires'
     return (rule' `Conditional.withCondition` unification')
+  where
+    removeRedundancy unificationSolution requires = do
+        let predicate =
+                (makePredicate `on` Predicate.toPredicate)
+                    unificationSolution requires
+        isPreRedundant <- evaluate predicate
+        let nonRedundantPredicate =
+                case isPreRedundant of
+                    Just True -> unificationSolution
+                    _         -> unificationSolution <> requires
+        Substitution.normalizeExcept nonRedundantPredicate
+
+    -- | ⌈L ∧ φ⌉ ∧ P → Pre
+    makePredicate unificationSolution requires =
```

Comment:
The comment helps, but the name `makePredicate` is a little vague: what kind of predicate does this make?